File: internal/emoji/emoji_test.go
package emoji

import "testing"

func TestNormalize(t *testing.T) {
	valid := []string{
		"👍",     // simple
		"👍🏽",    // skin tone modifier
		"👩‍💻",   // ZWJ sequence
		"🇳🇿",    // flag: two regional indicators
		"1️⃣",   // keycap
		"©",     // legacy pictographic
		"❤️",    // heart + VS16
		"🫱🏻‍🫲🏾", // handshake with mixed skin tones (ZWJ)
	}
	for _, s := range valid {
		if _, err := Normalize(s); err != nil {
			t.Errorf("Normalize(%q) rejected a valid emoji: %v", s, err)
		}
	}
	invalid := []string{
		"",
		"a",
		"hello",
		"👍👍",     // two clusters
		"👍 ",     // trailing space
		"🇳",      // lone regional indicator
		"@",      // punctuation
		"\u200d", // bare ZWJ
	}
	for _, s := range invalid {
		if got, err := Normalize(s); err == nil {
			t.Errorf("Normalize(%q) accepted an invalid reaction as %q", s, got)
		}
	}
	// Distinct keys: skin tones don't collide with the base emoji.
	a, _ := Normalize("👍")
	b, _ := Normalize("👍🏽")
	if a == b {
		t.Error("👍 and 👍🏽 collided")
	}
}

File: cfworker/test/emoji.test.ts
// Port of internal/emoji/emoji_test.go — every case verbatim. Runs on
// workerd, so Intl.Segmenter here is the same segmenter production uses.

import { describe, expect, it } from "vitest";
import { normalizeEmoji } from "../src/emoji";

describe("normalizeEmoji", () => {
  const valid = [
    "👍", // simple
    "👍🏽", // skin tone modifier
    "👩‍💻", // ZWJ sequence
    "🇳🇿", // flag: two regional indicators
    "1️⃣", // keycap
    "©", // legacy pictographic
    "❤️", // heart + VS16
    "🫱🏻‍🫲🏾", // handshake with mixed skin tones (ZWJ)
  ];
  it.each(valid)("accepts %s", (s) => {
    expect(normalizeEmoji(s)).not.toBeNull();
  });

  const invalid = [
    "",
    "a",
    "hello",
    "👍👍", // two clusters
    "👍 ", // trailing space
    "🇳", // lone regional indicator
    "@", // punctuation
    "‍", // bare ZWJ
  ];
  it.each(invalid)("rejects %j", (s) => {
    expect(normalizeEmoji(s)).toBeNull();
  });

  it("keeps skin tones distinct from the base emoji", () => {
    expect(normalizeEmoji("👍")).not.toBe(normalizeEmoji("👍🏽"));
  });
});
